universal-application-tool-0.0.1/app/auth/SamlCiviFormProfileAdapter.java

```diff
@@ -0,0 +1,172 @@
+    this.applicantRepositoryProvider = Preconditions.checkNotNull(applicantRepositoryProvider);
+    this.saml2Client = Preconditions.checkNotNull(client);
+    this.saml2Configuration = Preconditions.checkNotNull(configuration);
+  }
+
+  @Override
+  public Optional<UserProfile> create(
+      Credentials cred, WebContext context, SessionStore sessionStore) {
+    ProfileUtils profileUtils = new ProfileUtils(sessionStore, profileFactory);
+
+    Optional<UserProfile> samlProfile = super.create(cred, context, sessionStore);
+
+    if (samlProfile.isEmpty()) {
+      LOG.warn("Didn't get a valid profile back from SAML");
+      return Optional.empty();
+    }
+
+    if (!(samlProfile.get() instanceof SAML2Profile)) {
+      LOG.warn(
+          "Got a profile from SAML2 callback but it wasn't a SAML profile: %s", samlProfile.get());
+      return Optional.empty();
+    }
+
+    SAML2Profile profile = (SAML2Profile) samlProfile.get();
+
+    // Check if we already have a profile in the database for the user returned to us by SAML2
+    Optional<Applicant> existingApplicant =
+        applicantRepositoryProvider
+            .get()
+            .lookupApplicant(profile.getAttribute("email", String.class))
+            .toCompletableFuture()
+            .join();
+
+    // Now we have a three-way merge situation.  We might have
+    // 1) an applicant in the database (`existingApplicant`),
+    // 2) a guest profile in the browser cookie (`existingProfile`)
+    // 3) a SAML2Profile in the callback from the Identity Provider (`profile`).
+    // We will merge 1 and 2, if present, into `existingProfile`, then merge in `profile`.
+
+    Optional<CiviFormProfile> existingProfile = profileUtils.currentUserProfile(context);
+    if (existingApplicant.isPresent()) {
+      if (existingProfile.isEmpty()) {
+        // Easy merge case - we have an existing applicant, but no guest profile.
+        // This will be the most common.
+        existingProfile = Optional.of(profileFactory.wrap(existingApplicant.get()));
+      } else {
+        // Merge the two applicants and prefer the newer one.
+        // For account, use the existing account and ignore the guest account.
+        Applicant guestApplicant = existingProfile.get().getApplicant().join();
+        Account existingAccount = existingApplicant.get().getAccount();
+        Applicant mergedApplicant =
+            applicantRepositoryProvider
+                .get()
+                .mergeApplicants(guestApplicant, existingApplicant.get(), existingAccount)
+                .toCompletableFuture()
+                .join();
+        existingProfile = Optional.of(profileFactory.wrap(mergedApplicant));
+      }
+    }
+
+    // Now merge in the information sent to us by the SAML Identity Provider.
+    if (existingProfile.isEmpty()) {
+      LOG.debug("Found no existing profile in session cookie.");
+      return Optional.of(civiformProfileFromSamlProfile(profile));
+    } else {
```

Comment:
remove the else wrapping since it's the termination of the function when it's reached.

Reply:
fixed